Nets/PSMNet.py

```diff
@@ -0,0 +1,304 @@
+import tensorflow as tf
+import tensorflow.keras as keras
+from .utils import *
+import numpy as np
+
+from Nets import Stereo_net
+from Nets import sharedLayers
+from Data_utils import preprocessing
+
+class PSMNet(Stereo_net.StereoNet):
+    _valid_args = [
+        ("left_img", "meta op for left image batch"),
+        ("right_img", "meta op for right image batch"),
+        ("max_disp", "max_disp for the model")
+    ] + Stereo_net.StereoNet._valid_args
+    _netName = "PSMNet"
+
+    def __init__(self, **kwargs):
+        """
+        Creation of a PSMNet for stereo prediction
+        """
+        super(PSMNet, self).__init__(**kwargs)
+
+    def _validate_args(self, args):
+        """
+        Check that args contains everything that is needed
+        Valid Keys for args:
+            left_img: left image op
+            right_img: right image op
+        """
+        super(PSMNet, self)._validate_args(args)
+        if ("left_img" not in args) or ("right_img" not in args):
+            raise Exception('Missing input op for left and right images')
+        if 'max_disp' not in args:
+            print('WARNING: max_disp not setted, setting default value 192')
+            args['max_disp'] = 192
+        return args
+
+    def _preprocess_inputs(self, args):
+        self._left_input_batch = args['left_img']
+        self._restore_shape = tf.shape(args['left_img'])[1:3]
+        self._left_input_batch = tf.cast(self._left_input_batch, tf.float32)
+        self._left_input_batch = preprocessing.pad_image(
+            self._left_input_batch, 64)
+
+        self._right_input_batch = args['right_img']
+        self._right_input_batch = tf.cast(self._right_input_batch, tf.float32)
+        self._right_input_batch = preprocessing.pad_image(
+            self._right_input_batch, 64)
+        self.max_disp = args['max_disp']
+        self.image_size_tf = tf.shape(self._left_input_batch)[1:3]
+
+    def _build_network(self, args):
+        conv4_left = self.CNN(self._left_input_batch)
+        fusion_left = self.SPP(conv4_left)
+
+        conv4_right = self.CNN(self._right_input_batch, reuse=tf.compat.v1.AUTO_REUSE)
+        fusion_right = self.SPP(conv4_right, reuse=tf.compat.v1.AUTO_REUSE)
+
+        cost_vol = self.cost_vol(fusion_left, fusion_right, self.max_disp)
+        outputs = self.CNN3D(cost_vol)
+        disps = self.output(outputs)#size of (B, H, W),3out
+        self._disparities.append(disps[2])
+
+    def CNN(self, bottom, reuse=False):
+        with tf.compat.v1.variable_scope('CNN',reuse=reuse):
+            with tf.compat.v1.variable_scope('conv0',reuse=reuse):
+                if reuse == tf.compat.v1.AUTO_REUSE:
+                    tf.print(bottom, ['left_batch_size:',bottom])
+                self._add_to_layers('CNN/conv0/conv0_1', sharedLayers.conv2d(bottom, [3, 3, 3, 32], \
```

Comment:
you can set the shape of the filter to [3, 3, input_batch.get_shape()[-1].value, 32] to handle inputs with a number of channels different from 3

Reply:
line 70 has been changed to the following:

self._add_to_layers('CNN/conv0/conv0_1', sharedLayers.conv2d(bottom, [3, 3, bottom.get_shape()[-1].value, 32], 
                    activation=tf.nn.leaky_relu, batch_norm=True, apply_relu=True,strides=2, name='conv0_1',reuse=reuse))

The same error.